File: Tools/Postprocessing/F_Src/fboxinfo.f90
! print out information regarding the AMR levels and boxes

program fboxinfo

  use bl_space
  use bl_error_module
  use bl_constants_module
  use bl_IO_module
  use plotfile_module

  implicit none

  type(plotfile) pf
  integer :: unit
  integer :: i, j, f

  integer :: ncells, ncells_domain

  integer :: lo(MAX_SPACEDIM), hi(MAX_SPACEDIM)
  integer :: dim

  integer :: narg, farg
  character(len=256) :: fname

  logical :: full, gridfile, castro, levels

  unit = unit_new()

  full = .false.
  gridfile = .false.
  castro = .false.
  levels = .false.

  narg = command_argument_count()

  farg = 1
  do while ( farg <= narg )
     call get_command_argument(farg, value = fname)

     select case (fname)

     case ('-f', '--full')
        full = .true.

     case ('-g', '--gridfile')
         gridfile = .true.

     case ('-l', '--levels')
         levels = .true.

     case ('-c', '--castro')
         castro = .true.

     case default
        exit

     end select

     farg = farg + 1
  end do

  if (gridfile .and. full) then
     print *, "ERROR: cannot specify both full and gridfile modes"
     stop
  endif
  if (castro .and. full) then
     print *, "ERROR: cannot specify both full and castro modes"
     stop
  endif

  if (farg > narg) then
     print *, " "
     print *, "Dump out information about the AMR levels and boxes"
     print *, "Works with 1-, 2-, or 3-d datasets."
     print *, " "
     print *, "usage:"
     print *, "   fboxinfo [-f|--full] plotfile"
     print *, " "
     print *, "args:"
     print *, "   [-f|--full]     output detailed information about the boxes"
     print *, "   [-g|--gridfile] output a gridfile for use with test_average"
     print *, "   [-c|--castro]   output a gridfile for use with castro restart"
     print *, "   [-l|--levels]   just output the number of levels"
     print *, " "
     stop
  end if


  do f = farg, narg

     call get_command_argument(f, value = fname)
     call build(pf, fname, unit)

     if (.not. gridfile .and. .not. castro .and. .not. levels) then
        print *, "plotfile: ", trim(fname)
     end if

     dim = pf%dim


     ! output the number of boxes at each level
1000 format(1x,"level ", i3, ": number of boxes = ", i6, ", volume = ", f6.2, "%")
1001 format(1x,"      ", 1x, "  maximum zones =   ", i7)
1002 format(1x,"      ", 1x, "  maximum zones =   ", i7, " x ", i7)
1003 format(1x,"      ", 1x, "  maximum zones =   ", i7, " x ", i7, " x " i7)

     if (levels) then
        print *, pf%flevel
        cycle
     endif


     if (.not. gridfile .and. .not. castro) then
        do i = 1, pf%flevel
           ncells = 0
           do j = 1, nboxes(pf, i)
              ncells = ncells + volume(get_pbox(pf, i, j))
           enddo
           ncells_domain = volume(plotfile_get_pd_box(pf, i)) 

           write (*,1000) i, nboxes(pf, i), 100.0*dble(ncells)/ncells_domain

           lo = lwb(plotfile_get_pd_box(pf, i))
           hi = upb(plotfile_get_pd_box(pf, i))

           select case (dim)
           case (1)
              write (*,1001) hi(1)-lo(1)+1
              
           case (2)
              write (*,1002) hi(1)-lo(1)+1, hi(2)-lo(2)+1
              
           case (3)
              write (*,1003) hi(1)-lo(1)+1, hi(2)-lo(2)+1, hi(3)-lo(3)+1
              
           end select
           
           print *, " "
        enddo
     endif

2001 format(1x,"  box ", i5, ":  ("i5,")   ("i5,")")
2002 format(1x,"  box ", i5, ":  ("i5,",",i5,")   ("i5,",",i5,")")
2003 format(1x,"  box ", i5, ":  ("i5,",",i5,",",i5")   ("i5,",",i5,",",i5")")

     ! if desired, output detailed box infomation
     if (full) then

        do i = 1, pf%flevel
           
           print *, " "
           print *, " level ", i
              
           do j = 1, nboxes(pf, i)
              
              lo = lwb(get_box(pf, i, j))
              hi = upb(get_box(pf, i, j))
              
              select case (dim)
              case (1)
                 write (*,2001) j, lo(1), hi(1)
                 
              case (2)
                 write (*,2002) j, lo(1), lo(2), hi(1), hi(2)
                 
              case (3)
                 write (*,2003) j, lo(1), lo(2), lo(3), hi(1), hi(2), hi(3)
                 
              end select
              
           end do
        end do
        
     endif
     
2990 format(1x,i2)
     
2997 format(3x,"((",i5,              ") ("i5,") ( 0 ))  ", i5 )
2998 format(3x,"((",i5,",",i5,       ") ("i5,",",i5,") ( 0, 0 ))  ", i5)
2999 format(3x,"((",i5,",",i5,",",i5,") ("i5,",",i5,",",i5,") ( 0, 0, 0 ))  ", i5)
     
3001 format(6x,"((",i5,              ") ("i5,") ( 0 ))  ")
3002 format(6x,"((",i5,",",i5,       ") ("i5,",",i5,") ( 0, 0 ))  ")
3003 format(6x,"((",i5,",",i5,",",i5,") ("i5,",",i5,",",i5,") ( 0, 0, 0 ))  ")
     
     
     if (gridfile) then
        
        write (*,2990) pf%flevel
        
        do i = 1, pf%flevel
           
           lo = lwb(plotfile_get_pd_box(pf, i))
           hi = upb(plotfile_get_pd_box(pf, i))
           
           select case (dim)
           case (1)
              write (*,2997) lo(1), hi(1), nboxes(pf, i)
              
           case (2)
              write (*,2998) lo(1), lo(2), hi(1), hi(2), nboxes(pf, i)
              
           case (3)
              write (*,2999) lo(1), lo(2), lo(3), hi(1), hi(2), hi(3), nboxes(pf, i)
           end select
           
           do j = 1, nboxes(pf, i)
              
              lo = lwb(get_box(pf, i, j))
              hi = upb(get_box(pf, i, j))
              
              select case (dim)
              case (1)
                 write (*,3001) lo(1), hi(1)
                 
              case (2)
                 write (*,3002) lo(1), lo(2), hi(1), hi(2)
                 
              case (3)
                 write (*,3003) lo(1), lo(2), lo(3), hi(1), hi(2), hi(3)
                 
              end select
              
           end do
        end do
        
     endif
     
     if (castro) then
        
        if (pf%flevel .eq. 1) then
           print*, "ERROR: only use --castro option for multilevel"
           stop
        end if
        
        print*,pf%flevel-1
        
        do i = 2, pf%flevel
           
           print*, nboxes(pf, i)
           
           do j = 1, nboxes(pf, i)
              
              lo = lwb(get_box(pf, i, j))
              hi = upb(get_box(pf, i, j))
              
              select case (dim)
              case (1)
                 write (*,3001) lo(1)/2, (hi(1)-1)/2
                 
              case (2)
                 write (*,3002) lo(1)/2, lo(2)/2, (hi(1)-1)/2, (hi(2)-1)/2
                 
              case (3)
                 write (*,3003) lo(1)/2, lo(2)/2, lo(3)/2, (hi(1)-1)/2, (hi(2)-1)/2, (hi(3)-1)/2
                 
              end select
              
           end do
        end do
        
     endif
     
     
     print *, " "
     call destroy(pf)
     
  enddo

end program fboxinfo

File: Tools/Plotfile/fboxinfo.cpp
#include <AMReX.H>
#include <AMReX_Print.H>
#include <AMReX_PlotFileUtil.H>

using namespace amrex;

namespace {

}

void main_main()
{
    bool b_full = false;
    bool b_gridfile = false;
    bool b_levels = false;

    const int narg = amrex::command_argument_count();

    int farg = 1;
    while (farg <= narg) {
        const auto fname = get_command_argument(farg);
        if (fname == "-f" || fname == "--full") {
            b_full = true;
        } else if (fname == "-g" || fname == "--gridfile") {
            b_gridfile = true;
        } else if (fname == "-l" || fname == "--levels") {
            b_levels = true;
        } else {
            break;
        }
        ++farg;
    }

    if (b_gridfile && b_full) {
        amrex::Abort("ERROR: cannot specify both full and gridfile modes");
    }

    if (farg > narg) {
        amrex::Print() << "\n"
                       << " Dump out information about the AMR levels and boxes\n"
                       << " Works with 1-, 2-, or 3-d datasets.\n"
                       << "\n"
                       << " usage:\n"
                       << "    fboxinfo [-f|--full] plotfile\n"
                       << "\n"
                       << " args:\n"
                       << "    [-f|--full]     output detailed information about the boxes\n"
                       << "    [-g|--gridfile] output a gridfile for use with test_average\n"
                       << "    [-l|--levels]   just output the number of levels\n"
                       << '\n';
        return;
    }

    for (int f = farg; f <= narg; ++f) {
        const auto& fname = amrex::get_command_argument(f);
        PlotFileData plotfile(fname);

        if (!b_gridfile && !b_levels) {
            amrex::Print() << " plotfile: " << fname << "\n";
        }

        const int dim = plotfile.spaceDim();
        const int nlevels = plotfile.finestLevel()+1;

        if (b_levels) {
            amrex::Print() << " " << nlevels << '\n';
            continue;
        }

        if (!b_gridfile)
        {
            for (int ilev = 0; ilev < nlevels; ++ilev) {
                const Long nboxes = plotfile.boxArray(ilev).size();
                const Long ncells = plotfile.boxArray(ilev).numPts();
                const Box prob_domain = plotfile.probDomain(ilev);
                const auto ncells_domain = prob_domain.d_numPts();
                amrex::Print() << " level " << std::setw(3) << ilev
                               << ": number of boxes = " << std::setw(6) << nboxes
                               << ", volume = "
                               << std::fixed << std::setw(6) << std::setprecision(2)
                               << 100.*static_cast<double>(ncells)/ncells_domain << "%\n";
                if (dim == 1) {
                    amrex::Print() << "          maximum zones =   "
                                   << std::setw(7) << prob_domain.length(0) << "\n";
                } else if (dim == 2) {
                    amrex::Print() << "          maximum zones =   "
                                   << std::setw(7) << prob_domain.length(0)
                                   << " x "
                                   << std::setw(7) << prob_domain.length(1) << "\n";
                } else {
                    amrex::Print() << "          maximum zones =   "
                                   << std::setw(7) << prob_domain.length(0)
                                   << " x "
                                   << std::setw(7) << prob_domain.length(1)
                                   << " x "
                                   << std::setw(7) << prob_domain.length(2) << "\n";
                }
                amrex::Print() << '\n';
            }
        }

        if (b_full) {
            for (int ilev = 0; ilev < nlevels; ++ilev) {
                amrex::Print() << "\n  level " << ilev << "\n";
                const BoxArray& ba = plotfile.boxArray(ilev);
                const auto nboxes = static_cast<int>(ba.size());
                for (int ibox = 0; ibox < nboxes; ++ibox) {
                    const Box& b = ba[ibox];
                    if (dim == 1) {
                        amrex::Print() << "   box " << std::setw(5) << ibox
                                       << ":  (" << std::setw(5) << b.smallEnd(0)
                                       << ")"
                                       << "   (" << std::setw(5) << b.bigEnd(0)
                                       << ")"
                                       << "\n";
                    } else if (dim == 2) {
                        amrex::Print() << "   box " << std::setw(5) << ibox
                                       << ":  (" << std::setw(5) << b.smallEnd(0)
                                       <<    "," << std::setw(5) << b.smallEnd(1)
                                       << ")"
                                       << "   (" << std::setw(5) << b.bigEnd(0)
                                       <<    "," << std::setw(5) << b.bigEnd(1)
                                       << ")"
                                       << "\n";
                    } else {
                        amrex::Print() << "   box " << std::setw(5) << ibox
                                       << ":  (" << std::setw(5) << b.smallEnd(0)
                                       <<    "," << std::setw(5) << b.smallEnd(1)
                                       <<    "," << std::setw(5) << b.smallEnd(2)
                                       << ")"
                                       << "   (" << std::setw(5) << b.bigEnd(0)
                                       <<    "," << std::setw(5) << b.bigEnd(1)
                                       <<    "," << std::setw(5) << b.bigEnd(2)
                                       << ")"
                                       << "\n";
                    }
                }
            }
        }

        if (b_gridfile) {
            amrex::Print() << " " << std::setw(2) << nlevels << "\n";
            for (int ilev = 0; ilev < nlevels; ++ilev) {
                const BoxArray& ba = plotfile.boxArray(ilev);
                const Long nboxes = ba.size();
                const Box prob_domain = plotfile.probDomain(ilev);
                amrex::Print() << "   ";
                amrex::IntVect prob_domain_type = prob_domain.type();
                amrex::detail::box_write(amrex::Print(), prob_domain.smallEnd().begin(),
                                                         prob_domain.bigEnd().begin(),
                                                         prob_domain_type.begin(), dim);
                amrex::Print() << "  " << nboxes << "\n";
                for (int ibox = 0; ibox < nboxes; ++ibox) {
                    amrex::Print() << "      ";
                    amrex::IntVect ba_type = ba[ibox].type();
                    amrex::detail::box_write(amrex::Print(), ba[ibox].smallEnd().begin(),
                                                             ba[ibox].bigEnd().begin(),
                                                             ba_type.begin(), dim);
                    amrex::Print() << "\n";
                }
            }
        }

        amrex::Print() << '\n';
    }
}

int main (int argc, char* argv[])
{
    amrex::SetVerbose(0);
    amrex::Initialize(argc, argv, false);
    main_main();
    amrex::Finalize();
}
